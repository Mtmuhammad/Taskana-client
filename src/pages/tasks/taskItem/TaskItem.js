// an individual task to be rendered in task table

import { useState } from "react";
import useAxiosPrivate from "../../../hooks/useAxiosPrivate";
import useAuth from "../../../hooks/useAuth";
import EditModal from "../editTask/EditModal";
import DeleteModal from "../deleteTask/DeleteModal";

const TaskItem = ({
  setShowTasks,
  task,
  setErrMsg,
  filterTasks,
  setUserTasks,
  setSuccess,
}) => {
  const { auth } = useAuth();
  const axiosPrivate = useAxiosPrivate();
  const [checked, setChecked] = useState(
    task?.status === "Completed" ? true : false
  );

  // check  task. Set status to "completed".
  const handleCheck = async () => {
    let isMounted = true;
    const controller = new AbortController();
    try {
      await axiosPrivate.patch(
        `/tasks/${task?.id}`,

        JSON.stringify({ status: "Completed" }),
        {
          headers: { "Content-Type": "application/json" },
          withCredentials: true,
        }
      );

      const res = await axiosPrivate.get(`/tasks/${auth?.user?.empNumber}`, {
        signal: controller.signal,
      });
      isMounted && setUserTasks(res.data.tasks);
      setShowTasks(res.data.tasks);
      filterTasks(res);
      setChecked(!checked);
    } catch (err) {
      setErrMsg(err?.response?.data?.error?.message);
    }

    return () => {
      isMounted = false;
      controller.abort();
    };
  };

  // uncheck  task. Set status to "open"
  const handleUncheck = async () => {
    let isMounted = true;
    const controller = new AbortController();
    try {
      await axiosPrivate.patch(
        `/tasks/${task?.id}`,

        JSON.stringify({ status: "Open" }),
        {
          headers: { "Content-Type": "application/json" },
          withCredentials: true,
        }
      );

      const res = await axiosPrivate.get(`/tasks/${auth?.user?.empNumber}`, {
        signal: controller.signal,
      });
      isMounted && setUserTasks(res.data.tasks);
      setShowTasks(res.data.tasks);
      filterTasks(res);
      setChecked(!checked);
    } catch (err) {
      setErrMsg(err?.response?.data?.error?.message);
    }

    return () => {
      isMounted = false;
      controller.abort();
    };
  };

  return (
    <div
      data-testid="task-item"
      className="
       task-item
       p-3
       border-bottom
       position-relative
     "
    >
      <div className="d-flex align-items-start">
        <div className="w-100">
          <div
            className="
            
             d-flex
             align-items-start
             form-check
           "
          >
            {/* checkbox */}
            {checked ? (
              <input
                style={{ height: "1.5rem", width: "1.5rem" }}
                type="checkbox"
                className="form-check-input flex-shrink-0 me-3"
                id="checkbox1"
                onClick={handleUncheck}
                defaultChecked
              />
            ) : (
              <input
                style={{ height: "1.5rem", width: "1.5rem" }}
                type="checkbox"
                className="form-check-input flex-shrink-0 me-3"
                id="checkbox1"
                onClick={handleCheck}
                data-testid="task-checkbox"
              />
            )}
            <label className="form-check-label" htmlFor="checkbox1"></label>
            <div>
              <div>
                {/* task title */}
                <h5
                  data-testid="task-title"
                  className={checked ? "checked fs-4" : "fs-4"}
                >
                  {task?.title}{" "}
                  {task?.important && (
                    <i className="text-warning bx bxs-star"></i>
                  )}
                </h5>

                {/* task description */}
                <div
                  data-testid="task-description"
                  className={
                    checked ? "checked text-muted fs-3" : "text-muted fs-3"
                  }
                >
                  {task?.description}
                </div>

                {/* task date */}
                <span
                  data-testid="task-date"
                  className={
                    checked ? "checked fs-2 text-muted" : "fs-2 text-muted"
                  }
                >
                  <i className="bx bxs-calendar me-2"></i>
                  {task?.date}
                </span>
                {/* task btn group */}
                <div data-testid="task-btn-group" className="mt-3">
                  <button
                    type="button"
                    className="task-btn btn btn-warning text-light me-3"
                    data-bs-toggle="modal"
                    data-bs-target={`#editTask${task?.id}`}
                  >
                    Edit
                  </button>
                  <button
                    type="button"
                    data-bs-toggle="modal"
                    data-bs-target={`#deleteTask${task?.id}`}
                    className="task-btn btn btn-danger"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      {/* edit task modal */}
      <EditModal
        setUserTasks={setUserTasks}
        setShowTasks={setShowTasks}
        task={task}
        setErrMsg={setErrMsg}
        setSuccess={setSuccess}
      />
      {/* delete task modal */}
      <DeleteModal
        setUserTasks={setUserTasks}
        setShowTasks={setShowTasks}
        task={task}
        setErrMsg={setErrMsg}
        setSuccess={setSuccess}
        filterTasks={filterTasks}
      />
    </div>
  );
};

export default TaskItem;
