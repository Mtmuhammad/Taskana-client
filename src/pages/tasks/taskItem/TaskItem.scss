@use "../../../styles/abstracts/_variables" as *;
@use "../../../styles/abstracts/_mixins" as *;

.tasks {
  background-color: var(--primary-color-light);
  color: var(--text-color);
  padding: 2rem;
  border-radius: 0.5rem;
}

#search {
  font-size: 2rem;
  background-color: var(--primary-color-light);
  color: var(--text-color);
  padding: 1rem;
}

#task-search {
  background-color: var(--primary-color-light);
  color: var(--text-color);
  font-size: 1.5rem;
}

.task-filter {
  display: flex;
  padding: 1rem 1.5rem;
  margin-bottom: 1rem;
  align-items: center;
  border-radius: 1rem;
  transition: all ease 0.2s;
  background-color: var(--primary-color-light);
  color: var(--text-color);
  border: 1px solid var(--text-color);
  cursor: pointer;

  &:hover {
    background-color: var(--body-color);
  }

  &-count {
    line-height: none;
    font-size: 1rem;

    &.all {
      background-color: #ff505090 !important;
    }
    &.important {
      background-color: #ffc36d90 !important;
    }
    &.complete {
      background-color: #00d09c90 !important;
    }
  }

  &-icon {
    display: flex;
    font-size: 2rem;
  }

  &-title {
    margin-bottom: 0;
    font-size: 1.5rem;
    padding-left: 0.5rem;
    text-transform: capitalize;
  }
}

.task-active {
  background-color: var(--body-color);
}

.form-check-input:checked {
  background-color: $primary-green;
  border-color: $primary-green;
}

.task-btn {
  font-size: 1.5rem;
}

.checked {
  text-decoration: line-through;
}
