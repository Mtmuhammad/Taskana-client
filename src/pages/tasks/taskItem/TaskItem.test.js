import React from "react";
import { render, screen } from "@testing-library/react";
import TaskItem from "./TaskItem";
import { BrowserRouter } from "react-router-dom";

describe("smoke and snapshot tests", () => {
  test("should render without crashing", () => {
    render(<TaskItem />, { wrapper: BrowserRouter });
  });

  test("should match snapshot", () => {
    const { asFragment } = render(<TaskItem />, { wrapper: BrowserRouter });
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("task item", () => {
  test("should render task item container", () => {
    render(<TaskItem />, { wrapper: BrowserRouter });
    const container = screen.getByTestId("task-item");
    expect(container).toBeInTheDocument();
    expect(container).toHaveClass(
      "task-item p-3 border-bottom position-relative"
    );
  });
  test("should render task checkbox", () => {
    render(<TaskItem />, { wrapper: BrowserRouter });
    const checkbox = screen.getByTestId("task-checkbox");
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).toHaveClass("form-check-input flex-shrink-0 me-3");
  });
  test("should render task title", () => {
    render(<TaskItem />, { wrapper: BrowserRouter });
    const title = screen.getByTestId("task-title");
    expect(title).toBeInTheDocument();
    expect(title).toHaveClass("fs-4");
  });
  test("should render task description", () => {
    render(<TaskItem />, { wrapper: BrowserRouter });
    const description = screen.getByTestId("task-description");
    expect(description).toBeInTheDocument();
    expect(description).toHaveClass("fs-3");
  });
  test("should render task date", () => {
    render(<TaskItem />, { wrapper: BrowserRouter });
    const date = screen.getByTestId("task-date");
    expect(date).toBeInTheDocument();
    expect(date).toHaveClass("fs-2 text-muted");
  });
  test("should render task btn group", () => {
    render(<TaskItem />, { wrapper: BrowserRouter });
    const btnGroup = screen.getByTestId("task-btn-group");
    expect(btnGroup).toBeInTheDocument();
    expect(btnGroup).toHaveClass("mt-3");

    const buttons = screen.getAllByRole("button");
    expect(buttons.length).toBe(2);
  });
});
