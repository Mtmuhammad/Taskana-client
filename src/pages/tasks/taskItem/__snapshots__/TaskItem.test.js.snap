// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="
 task-item
 p-3
 border-bottom
 position-relative
 "
    data-testid="task-item"
  >
    <div
      class="d-flex align-items-start"
    >
      <div
        class="w-100"
      >
        <div
          class="
 d-flex
 align-items-start
 form-check
 "
        >
          <input
            class="form-check-input flex-shrink-0 me-3"
            data-testid="task-checkbox"
            id="checkbox1"
            style="height: 1.5rem; width: 1.5rem;"
            type="checkbox"
          />
          <label
            class="form-check-label"
            for="checkbox1"
          />
          <div>
            <div>
              <h5
                class="fs-4"
                data-testid="task-title"
              >
                 
              </h5>
              <div
                class="text-muted fs-3"
                data-testid="task-description"
              />
              <span
                class="fs-2 text-muted"
                data-testid="task-date"
              >
                <i
                  class="bx bxs-calendar me-2"
                />
              </span>
              <div
                class="mt-3"
                data-testid="task-btn-group"
              >
                <button
                  class="task-btn btn btn-warning text-light me-3"
                  data-bs-target="#editTaskundefined"
                  data-bs-toggle="modal"
                  type="button"
                >
                  Edit
                </button>
                <button
                  class="task-btn btn btn-danger"
                  data-bs-target="#deleteTaskundefined"
                  data-bs-toggle="modal"
                  type="button"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="modal fade"
      data-testid="edit-modal"
      id="editTaskundefined"
      style="color: black;"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-md modal-dialog-scrollable"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title  fw-bold"
              data-testid="header"
              id="editTaskLabel"
            >
               Edit Task
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              data-bs-dismiss="modal"
              data-testid="close-btn"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="name"
              >
                Task Title
              </label>
              <input
                class="form-control"
                data-testid="input"
                id="title"
                name="title"
                placeholder="What is the title of this task?"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="createTaskDescription"
              >
                Description (optional)
              </label>
              <textarea
                class="form-control"
                data-testid="input"
                id="createTaskDescription"
                name="description"
                placeholder="Add any extra details about this task"
                rows="3"
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                for="createTaskImportant"
              >
                Is this task urgent?
              </label>
              <select
                class="form-control"
                data-testid="input"
                id="createTaskImportant"
                name="important"
                style="font-size: 1.3rem;"
              >
                <option
                  default=""
                  value=""
                >
                  Choose an option:
                </option>
                <option
                  value="true"
                >
                  Yes
                </option>
                <option
                  value="false"
                >
                  No
                </option>
              </select>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="fs-5 btn btn-secondary"
              data-bs-dismiss="modal"
              data-testid="cancel-btn"
              type="button"
            >
              Cancel
            </button>
            <button
              class="fs-5 btn btn-success"
              data-bs-dismiss="modal"
              data-testid="edit-btn"
              type="button"
            >
              Edit
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="modal fade"
      data-testid="delete-modal"
      id="deleteTaskundefined"
      style="display: none; color: black;"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-md modal-dialog-scrollable"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title  fw-bold"
              data-testid="header"
              id="deleteProjectLabel"
            >
               Delete item Permanently?
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              data-bs-dismiss="modal"
              data-testid="close-btn"
              type="button"
            />
          </div>
          <div
            class="modal-body justify-content-center flex-column d-flex"
          >
            <i
              class="bx bxs-trash text-danger display-2 text-center mt-2"
              data-testid="delete-icon"
            />
            <p
              class="mt-4 fs-3 text-center"
              data-testid="delete-text"
            >
              You can only delete this item Permanently
            </p>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="fs-5 btn btn-secondary"
              data-bs-dismiss="modal"
              data-testid="cancel-btn"
              type="button"
            >
              Cancel
            </button>
            <button
              class="fs-5 btn btn-danger color-fff"
              data-bs-dismiss="modal"
              data-testid="delete-btn"
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
